src/storage/mvcc/txn.rs

```diff
@@ -253,13 +272,22 @@ impl<S: Snapshot, P: PdClient + 'static> MvccTxn<S, P> {
                 "async commit is not yet compatible with large transactions"
             );
 
-            // TODO(nrc) this is going to block all the other keys' processing and writing, we should
-            // do it async.
-            // TODO(nrc) this is also unsound! If we don't complete taking the lock until after another
-            // node gets a start ts to read the key, it can violate the snapshot property.
-            let ts = ::futures_executor::block_on(Compat01As03::new(self.pd_client.get_tso()))?;
-            lock.min_commit_ts = ts;
+            // This operation should not block because the latch makes sure only one thread
+            // is operating on this key.
+            let key_guard =
+                CONCURRENCY_MANAGER_LOCK_DURATION_HISTOGRAM.observe_closure_duration(|| {
+                    ::futures_executor::block_on(self.concurrency_manager.lock_key(&key))
+                });
+
+            let ts = key_guard.with_lock(|l| {
+                let ts = self.concurrency_manager.max_read_ts().next();
```

Comment:
Is it possible the `ts` is less than the `start_ts` or safe point?

Reply:
Good catch. It must be at least `start_ts + 1`.

Reply:
I think the safe point cannot be smaller than `start_ts`.

Reply:
The min_commit_ts of prewrite is either `start_ts + 1` or `for_update_ts + 1`. https://github.com/pingcap/tidb/blob/master/store/tikv/prewrite.go#L57 
Can we just use max(`min_commit_ts`, `max_read_ts + 1`) as the lock's `min_commit_ts`?

Reply:
Yes.

Reply:
I was wrong... TiDB sets min_commit_ts to 0 if using async commit because semantically a positive min_commit_ts means the large txn protocol is enabled.

So I set it to `max(start_ts, for_update_ts, max_read_ts) + 1`.